piff/gsobject_model.py

```diff
@@ -0,0 +1,389 @@
+from .util import hsm
+
+
+class GSObjectModel(Model):
+    """ Model that takes a fiducial GalSim.GSObject and dilates, shifts, and shears it to get a
+    good match to stars.
+
+    :param gsobj:    GSObject to use as fiducial profile.
+    :param fastfit:  Use HSM moments for fitting.  Approximate, but fast.  [default: False]
+    :param force_model_center: If True, PSF model centroid is fixed at origin and
+                        PSF fitting will marginalize over stellar position.  If False, stellar
+                        position is fixed at input value and the fitted PSF may be off-center.
+                        [default: True]
+    :param include_pixel:   Include integration over pixel?  [default: True]
+    :param logger:   A logger object for logging debug info. [default: None]
+    """
+    def __init__(self, gsobj, fastfit=False, force_model_center=True, include_pixel=True,
+                 logger=None):
+        import galsim
+        if isinstance(gsobj, basestring):
+            gsobj = eval(gsobj)
+
+        self.kwargs = {'gsobj':repr(gsobj),
+                       'fastfit':fastfit,
+                       'force_model_center':force_model_center,
+                       'include_pixel':include_pixel}
+
+        self.gsobj = gsobj
+        self._fastfit = fastfit
+        self._force_model_center = force_model_center
+        self._method = 'auto' if include_pixel else 'no_pixel'
+        # Params are [u, v], size, g1, g2
+        if self._force_model_center:
+            self._nparams = 3
+            params = [1.0, 0.0, 0.0]
+        else:
+            self._nparams = 5
+            params = [0.0, 0.0, 1.0, 0.0, 0.0]
+
+        # Calibrate gsobj by measuring it with HSM.  This way we can use differences in HSM moments
+        # to get a reasonable starting guess for stars.
+        prof = self.getProfile(params)
+        img = prof.drawImage(method=self._method)
+        sd = StarData(img, img.trueCenter())
+        fiducial_star = Star(sd, None)
+        flux, cenu, cenv, size, g1, g2, flag = hsm(fiducial_star)
+        if flag != 0:
+            raise RuntimeError("Error calculating fiducial moments for this gsobject.")
+        self._fiducial_size = size
+        self._fiducial_shape = galsim.Shear(g1=g1, g2=g2)
+        self._fiducial_flux = flux
+        self._fiducial_centroid = galsim.PositionD(cenu, cenv)
+
+    def moment_fit(self, star, logger=None):
+        """Estimate transformations needed to bring self.gsobj towards given star."""
+        import galsim
+        flux, cenu, cenv, size, g1, g2, flag = hsm(star)
+        centroid = galsim.PositionD(cenu, cenv)
+        shape = galsim.Shear(g1=g1, g2=g2)
+
+        if flag:
+            raise RuntimeError("Failed measuring HSM moments of star.")
+
+        dflux = flux/self._fiducial_flux
+        dcentroid = centroid - self._fiducial_centroid
+        dsize = size/self._fiducial_size
+        dshape = shape - self._fiducial_shape
```

Comment:
I see.  I think maybe I have a little better understanding of how this fiducial stuff works now.  (I was confused earlier about what it was intended for -- mostly because I only looked at the parameters for the Gaussian case, where things are all supposed to be either 1 or 0.  And for Gaussian, it's only close but not perfect because of the GalSim automatic pixel scale bit.

So, I think we probably do want to bump up the GalSim accuracy slightly then.  Probably maxk_threshold=folding_threshold=1.e-4 would be appropriate.  And I don't think that would slow things down all that much for these nicely smooth profiles.  In particular, maxk_threshold is the one that mostly governs the pixel scale in real space.